### ISSUE TITLE
Define a standard for the composer.json file

### ISSUE THREAD

**Issue by Soullivaneuh** [composer/composer#5494]
The `composer.json` can have a different formats, especially for the version constraint.

Some are recommended, other deprecated. The doc shows examples with the latest ones, but does not clearly define what it SHOULD be used.

We should do this.

I was thinking about:
- Indent: 4 spaces indentation
- OR constraint: Should be `||` (and not `||` or `|`)
- Caret operator using instead of tilde (`~2.3,>=2.3.3` -> `^2.3.3`)
- Precise that PHP should have the same rules about [unbound constraint](https://getcomposer.org/doc/faqs/why-are-unbound-version-constraints-a-bad-idea.md). cc @Ocramius 
- Ordered packages?
- Sections order (`title` before `description` etc etc...)?

Anything else?

**Comment by Soullivaneuh** [comment]
@Ocramius: Already proposed in https://github.com/composer/composer/issues/5257 but refused for other reasons.

This is why I wrote this plugin: https://github.com/Soullivaneuh/composer-lint

But nothing official. This is why I would like to clarify things on the documentation to make everyone "agree".

**Comment by greg0ire** [comment]
:+1: , this is more a documentation issue, `composer validate` should be here to enforce what can be enforced, but first a standard should be defined, ideally.

**Comment by stof** [comment]
``composer validate`` should accept any valid config. It is not about checking coding standards in the file (using 2 spaces for the indentation does not make the file invalid for usage in composer, while using a broken constraint does). Adding anything related to coding standards in ``composer validate`` would make it broken (even if we provide an official coding standard, anyone would be fine if they define their own coding standard for the composer.json files in their projects)

**Comment by greg0ire** [comment]
Maybe there could be `strict` mode? Or a separate `composer lint` command just for that ? Also `composer validate` has several levels apparently, if it issues a small notice for a valid file, I guess you can still say it "accepts" it.

**Comment by stof** [comment]
@greg0ire matching coding standards or no does not make the file invalid. A coding standard checker should be a separate tool.

**Comment by greg0ire** [comment]
> @greg0ire matching coding standards or no does not make the file invalid. A coding standard checker should be a separate tool.

Then I guess some warnings currently issued by `composer validate` should be moved to this tool. Things like : 

> Defining autoload.psr-4 with an empty namespace prefix is a bad idea for performance

don't make the Composer manifest invalid IMO.

**Comment by slbmeh** [comment]
That warning is probably better suited for the autoload generator.

**Comment by alcohol** [comment]
Just to be clear, the caret and tilde operator are not completely identical in behaviour. So blindly replacing one with the other could lead to unintended results.

**Comment by alcohol** [comment]
TLDR; I am very much against enforcing any of these "rules".

---

> Ordered packages?

This does not have to be enforced in my humble opinion. It serves no real purpose. Also, `require` has the `--sort-packages` flag should users have a desire to sort their packages.

> Sections order (title before description etc etc...)?

Again, serves no real purpose.

> Indent: 4 spaces indentation

I really do not see a reason or need to differentiate in the configuration file between tabs vs spaces, 2 vs 4, etc. It is just whitespace.

> Precise that PHP should have the same rules about unbound constraint.

I am not quite sure what you mean with this sentence. Language barrier probably. Can you rephrase please?

> OR constraint

We already hint at using `||` everywhere (we replaced all occurrences of `|` in the documentation and Composer's own `composer.json`). Whether or not you use whitespace around is really not important.

**Comment by greg0ire** [comment]
I think there is a real purpose to all this rules : limiting git conflicts. The Composer manifest of a project is one of the most edited file, and having a deterministic output for it should make conflicts less frequent : for instance, if people always add packages at the end of the `require` section as opposed to sorting them alphabetically, PRs will conflict very often.

> I am not quite sure what you mean with this sentence. Language barrier probably. Can you rephrase please?

I think @Soullivaneuh means that php makes no exception to the "don't use unbound constraints" rule.

**Comment by Soullivaneuh** [comment]
> Just to be clear, the caret and tilde operator are not completely identical in behaviour. So blindly replacing one with the other could lead to unintended results.

Agree. But, stop me if I'm wrong, the caret was introduce because the tilde does not fully respect semver, especially for 0.x releases. So it should be recommended to use them instead IMHO.

> I am very much against enforcing any of these "rules".

I never said something like this! :wink: 

The issue is about writing a convention. A kind of Best Practices like it was done for Symfony.

Of course the developer can follow this, or not. It's a choice, like for PSR standards.

> > Precise that PHP should have the same rules about unbound constraint.
> I am not quite sure what you mean with this sentence. Language barrier probably. Can you rephrase please?

Yeah, certainly my bad English, agains. :-)

I was talking about this: https://getcomposer.org/doc/faqs/why-are-unbound-version-constraints-a-bad-idea.md

A note should be added for PHP and HHVM constraints specify it should not be unbound either.

> > OR constraint
> We already hint at using || everywhere (we replaced all occurrences of | in the documentation and Composer's own composer.json). Whether or not you use whitespace around is really not important.

On the documentation and examples yes, but no convention rule about that.

In this case we can say the same for the 4 spaces indent. You use 4 spaces on our docs, so it should be followed too.

**Comment by alcohol** [comment]
But should this not be up to individual projects? Composer does not care about how projects format their `composer.json`. We only have (a minimal set of) guidelines for our own codebase. What kind of guidelines other projects want to impose or suggest, should really be their prerogative in my humble opinion.

**Comment by greg0ire** [comment]
> What kind of guidelines other projects want to impose or suggest, should really be their prerogative in my humble opinion.

That's right, but I think you could recommend a way that you think works best for various reasons. This would help limiting pedantic debates on each project about what convention to adopt, once the debates in question have taken place here.

**Comment by Soullivaneuh** [comment]
> should really be their prerogative in my humble opinion.

This is the same PHP standards in this case, right?

Again, as @greg0ire says, I just would like to have a convention. After that, people can follow them, or not. :wink:

**Comment by Garbee** [comment]
You have your own package already to lint things the way you want. I don't see why composer itself needs to get into caring about this stuff.

On the note of git conflicts, that is up to projects to determine the best way to handle for themselves. Why should everyone conform to needing to alphabetize their lists when they get along just fine without doing so since dependencies are added so few times?

Everything being suggested here is *personal preference* and composer as the base should not care. Just like PHP doesn't care about how many classes are in a file or if you use tabs vs spaces.

Overall, the suggestions here do *not impact developers at all* on a _regular basis_. Therefore, why should anyone care? It seems all we'd end up doing is water-shedding over the details, implement it, then only have a minor subset of users using it. And then it would need to be maintained as well in the future. This should be on external linters to handle on a project-by-project basis for their needs, not built into core. This way, projects can use whatever convention and linter they want and the ecosystem as a whole isn't affected one bit by it.

**Comment by greg0ire** [comment]
> Why should everyone conform to needing to alphabetize their lists when they get along just fine without doing so since dependencies are added so few times?

Because adding packages at the end also makes the diff ugly because there is no trailing comma after the last element. The diff will look like this : 

```diff
- existing/package: "^1.0"
+ existing/package: "^1.0",
+ new/package: "^1.0",
```

Who likes to review that?

> Everything being suggested here is personal preference and composer as the base should not care. Just like PHP doesn't care about how many classes are in a file or if you use tabs vs spaces.

Still, isn't there a standard about precisely that tabs vs spaces and number of classes in a file?

> This should be on external linters to handle on a project-by-project basis for their needs

This is precisely what we want avoid here: having to think about that on a project-by-project basis. Just having to say "we follow the Composer standard" and be done with it would be great.

**Comment by slbmeh** [comment]
> Because adding packages at the end also makes the diff ugly because there is no trailing comma after the last element.

You still end up with the diff problem when the package is last alphabetically.  When putting them in order by addition you can use a leading comma instead of a trailing comma.

```json
{
    "require": {
        "php": "^5.5.9 || ^7.0"
        ,"symfony/symfony": "^3.0"
    }
}
```

> Still, isn't there a standard about precisely that tabs vs spaces and number of classes in a file?

The tabs vs spaces is a rule in a coding standard maintained by a separate standards organization.

I don't feel it makes sense for composer to manage a standard. PHP does not maintain a standard for how to utilize it.

**Comment by greg0ire** [comment]
> You still end up with the diff problem when the package is last alphabetically.

Yeah, but the probability of this happening is `1/n`, n being the number of packages including the one being added, so definitely an improvement.

> I don't feel it makes sense for composer to manage a standard. PHP does not maintain a standard for how to utilize it.

That's a good point, but sadly, I think it is unlikely that an organization would be created just for that… and I don't think PSR will nor should do it.

**Comment by Soullivaneuh** [comment]
> Therefore, why should anyone care?

And why we should not?

This is indeed to the user/organization to follow a standard or not. This is a recommendation, not an obligation.

Some developers like to follow a standard too (yes, they exist!) and for this case no standard is available.

Maybe I'm missing something, but I don't really see the hurt to add a standard if some peoples don't want to follow it...

For the rest, @greg0ire already answered. :wink:

**Comment by Garbee** [comment]
This kind of "standard" will only end up in numerous wasted man-hours on the part of the Composer project and anyone involved in deciding what the rules should be.

Best Composer kick it out for people to do on their own (as you already are) and be done with it. Composer doesn't have anything to worry about here internally, so it doesn't gain from any kind of linting rules.

P.S. 4 spaces for just the JSON is a wonky move in the first place. Many JS standards are 2-spaced indent which means you'd have *one* file where it would need to be 4 spaced. That is very chaotic when you want consistent code style. So you already need to explain why using 4 spaces is better than 2 for the JSON. This is just *one* point we'd have endless rounds about that won't go anywhere.

**Comment by Soullivaneuh** [comment]
> to do on their own (as you already are)

I'm not already doing this. I'm trying to follow as possible sample from the doc. This is especially why I would like to have an official guide for that.

> P.S. 4 spaces for just the JSON is a wonky move in the first place. Many JS standards are 2-spaced indent which means you'd have one file where it would need to be 4 spaced. That is very chaotic when you want consistent code style. So you already need to explain why using 4 spaces is better than 2 for the JSON. This is just one point we'd have endless rounds about that won't go anywhere.

Quite agree. As I said I follow samples. :wink: 

But if 2 spaces seems more logic, and it seems to, we should consider it.

**Comment by greg0ire** [comment]
> This kind of "standard" will only end up in numerous wasted man-hours on the part of the Composer project and anyone involved in deciding what the rules should be.

I think we could sum up our point of view by saying that in our opinion, those man-hours wouldn't be wasted.

> 4 spaces for just the JSON is a wonky move in the first place. Many JS standards are 2-spaced indent which means you'd have one file where it would need to be 4 spaced. That is very chaotic when you want consistent code style. So you already need to explain why using 4 spaces is better than 2 for the JSON. This is just one point we'd have endless rounds about that won't go anywhere.

For instance, you're making a very good point here, and I hope you can see how having this discussion take place here would make it an authoritative reference . Having it discussed once and for all would allow the project to give a sane recommendation, with a public discussion behind, thus sparing these man-hours to projects that use Composer. 
It's kind of a time sacrifice if you will, and if the Composer team does not want to do it, which is completely understandable and ok, let them close this issue.

**Comment by greg0ire** [comment]
I must admit this particular point about conflicting standards makes a some sense, what do you think @Soullivaneuh

**Comment by Soullivaneuh** [comment]
> I must admit this particular point about conflicting standards makes a some sense, what do you think @soullivaneuh?

Yes and no.

First @Garbee, your case is concerning JSON and not global composer rules . Yeah it would be cool to have a JSON standard and each rely on it, would be the best. But I don't know if it exist.

But this have nothing to do with the caret rule or the `OR` constraint, this is specific to composer.

And finally, I don't think npn and bower rely on the `composer.json` file but on they owns. So I don't really see the problem here.

The only issue here would be about the JSON format AFAIK, so about tab, 2 spaces or 4 spaces. This is only a tiny part of the issue here.

**Comment by Garbee** [comment]
The problem of course isn't about the *file* but the *file type* by which linters execute against. You'd end up with multiple package json files being excluded from your project's check unless all of them somehow magically align.

And the spacing is actually a part of the fundamental concept being discussed, so it is far from minor. It is just one glaring hole in how this proposal makes little sense for any upstream project to care about.

**Comment by greg0ire** [comment]
@Soullivaneuh is quite right in saying that we don't want to validate the format as much as the contents. After all, version constraints could be expressed in XML as well, like this : `<version>^2.0 | 3.0</version>`. Different format, same contents, and same rules. So instead of validating a file, maybe the file should be parsed in an object, and what would be validated would be that object. No ESLint involved.

To better illustrate the problem, here is [the kind of discussion](https://github.com/composer/composer/pull/4910) that isn't solved because there is no standard, and spreads to [symfony](https://github.com/symfony/symfony/pull/18186), and then in [DunglasActionBundle](https://github.com/dunglas/DunglasActionBundle/pull/51), and maybe many others, I don't know. Lots of noise in many projects b/c if there are two ways of doing one thing, which one should we use, and if it does not matter, why are there two ways?

BTW, the point to not recommend `||` over `|` here is that you would get warnings from external projects. Surely it could be possible to validate only the current manifest instead of the manifest of all dependencies?
